src/common/loki.cpp

```diff
@@ -394,3 +394,157 @@ loki_exp2 (double x)
+     each addition step and before the following comparison at the caller's
+     site.  It is necessary on x86 systems where double-floats are not IEEE
+     compliant by default, to avoid that the results become platform and
+     compiler option dependent.  'volatile' is a portable alternative to gcc's
+     -ffloat-store option.  */
+  volatile double y = x;
+  volatile double z = y;
+
+  if (z > 0.0)
+    {
+      /* Avoid rounding error for x = 0.5 - 2^(-DBL_MANT_DIG-1).  */
+      if (z < 0.5)
+        z = 0.0;
+      /* Avoid rounding errors for values near 2^k, where k >= DBL_MANT_DIG-1.  */
+      else if (z < TWO_MANT_DIG)
+        {
+          /* Add 0.5 to the absolute value.  */
+          y = z += 0.5;
+          /* Round to the next integer (nearest or up or down, doesn't
+             matter).  */
+          z += TWO_MANT_DIG;
+          z -= TWO_MANT_DIG;
+          /* Enforce rounding down.  */
+          if (z > y)
+            z -= 1.0;
+        }
+    }
+  else if (z < 0.0)
+    {
+      /* Avoid rounding error for x = -(0.5 - 2^(-DBL_MANT_DIG-1)).  */
+      if (z > - 0.5)
+        z = MINUS_ZERO;
+      /* Avoid rounding errors for values near -2^k, where k >= DBL_MANT_DIG-1.  */
+      else if (z > -TWO_MANT_DIG)
+        {
+          /* Add 0.5 to the absolute value.  */
+          y = z -= 0.5;
+          /* Round to the next integer (nearest or up or down, doesn't
+             matter).  */
+          z -= TWO_MANT_DIG;
+          z += TWO_MANT_DIG;
+          /* Enforce rounding up.  */
+          if (z < y)
+            z += 1.0;
+        }
+    }
+  return z;
+}
+
+// adapted from Lokinet llarp/encode.hpp
+// from  https://en.wikipedia.org/wiki/Base32#z-base-32
+static const char zbase32_alpha[] = {'y', 'b', 'n', 'd', 'r', 'f', 'g', '8',
+                                     'e', 'j', 'k', 'm', 'c', 'p', 'q', 'x',
+                                     'o', 't', '1', 'u', 'w', 'i', 's', 'z',
+                                     'a', '3', '4', '5', 'h', '7', '6', '9'};
+
+/// adapted from i2pd
+template <typename v, typename stack_t>
+const char* base32z_encode(const v& value, stack_t &stack)
+{
+  size_t ret = 0, pos = 1;
+  int bits = 8, tmp = value[0];
+  size_t len = value.size();
+  while(ret < sizeof(stack) && (bits > 0 || pos < len))
+  {
+    if(bits < 5)
+    {
+      if(pos < len)
+      {
+        tmp <<= 8;
+        tmp |= value[pos] & 0xFF;
+        pos++;
+        bits += 8;
+      }
+      else  // last byte
+      {
+        tmp <<= (5 - bits);
+        bits = 5;
+      }
+    }
+
+    bits -= 5;
+    int ind = (tmp >> bits) & 0x1F;
+    if(ret < sizeof(stack))
+    {
+      stack[ret] = zbase32_alpha[ind];
+      ret++;
+    }
+    else
+      return nullptr;
+  }
+  return &stack[0];
+}
+
+std::string loki::hex64_to_base32z(const std::string &src)
+{
+  char buf[128]      = {};
+  std::string result = base32z_encode(src, buf);
```

Comment:
Do we need to handle the case where `base32z_encode` returns a `nullptr`?

Reply:
Also, kind of annoying that we throw away the buffer, can we use the buffer inside the string being returned instead?

Reply:
Added a nullptr check. string.data() returns const buffer and seem iffy about const casting that out.